rust-escrow/stable-escrow/src/escrow.rs

```diff
@@ -0,0 +1,482 @@
+     * Transfer self NEP141 of the stable NEP141 amount as a receipt
+     */
+    #[private]
+    pub fn deposit(&mut self, sender_id: AccountId, amount: Balance) {
+        if !self.is_deposit_allowed() {
+            env::panic_str("ERR_DEPOSIT_NOT_ALLOWED");
+        }
+
+        if amount > self.get_metadata().unpaid_amount {
+            env::panic_str("ERR_DEPOSIT_NOT_ALLOWED");
+        }
+
+        // Fee Calculations
+        let fee_amount = (amount as f32 * self.get_fees().percentage) as Balance;
+        self.fees.balance = self
+            .get_fees()
+            .balance
+            .checked_add(fee_amount)
+            .unwrap_or_else(|| env::panic_str("ERR_FEE_BALANCE_OVERFLOW"));
+        let amount_minus_fee = amount
+            .checked_sub(fee_amount)
+            .unwrap_or_else(|| env::panic_str("ERR_AMOUNT_OVERFLOW"));
+
+        // Register transfer
+        self.ft.internal_deposit(&sender_id, amount_minus_fee);
+        self.deposits.insert(&sender_id);
+        self.get_metadata().unpaid_amount = self
+            .metadata
+            .unpaid_amount
+            .checked_sub(amount)
+            .unwrap_or_else(|| env::panic_str("ERR_UNPAID_AMOUNT_OVERFLOW"));
+
+        log!(
+            "[deposit]: sender_id: {}, amount_minus_fee: {}, fee: {}",
+            sender_id,
+            amount_minus_fee,
+            fee_amount
+        );
+    }
+
+    /**
+     * Only if total funds are not reached or escrow has expired
+     * Transfer all funds to receiver_id
+     */
+    #[payable]
+    pub fn withdraw(&mut self) -> Promise {
+        if !self.is_withdrawal_allowed() {
+            env::panic_str("ERR_WITHDRAWAL_NOT_ALLOWED");
+        }
+
+        let receiver_id = env::signer_account_id();
+        let amount = self.ft.internal_unwrap_balance_of(&receiver_id);
+
+        let promise = Promise::new(self.get_metadata().nep_141.clone()).function_call(
+            "ft_transfer".to_string(),
+            json!({
+                "amount": amount.to_string(),
+                "receiver_id": receiver_id.to_string(),
+            })
+            .to_string()
+            .into_bytes(),
+            1, // 1 yoctoNEAR
+            GAS_FT_TRANSFER,
+        );
+
+        let callback = Promise::new(env::current_account_id()).function_call(
+            "on_withdraw_callback".to_string(),
+            json!({
+                "receiver_id": receiver_id.to_string(),
+                "amount": amount.to_string(),
+            })
+            .to_string()
+            .into_bytes(),
+            0,
+            GAS_FT_TRANSFER_CB,
+        );
+
+        promise.then(callback)
+    }
+
+    #[payable]
+    pub fn claim_fees(&mut self) -> Promise {
+        if !self.is_withdrawal_allowed() {
+            if self.is_deposit_allowed() || self.is_withdrawal_allowed() {
+                env::panic_str("ERR_CLAIM_FEES_NOT_ALLOWED");
+            }
+        }
```

Comment:
@susguzman this logic seems smelly to me. First we check if it's not allowed, but then if it's allowed?

Reply:
Agree. What I am trying to do is enable the claim only when the Escrow is expired or successful. Any ideas ?

Reply:
New logic added !